nixos/modules/services/network-filesystems/ipfs.nix

```diff
@@ -236,7 +242,31 @@ in
 
     systemd.packages = [ cfg.package ];
 
-    systemd.services.ipfs = {
+    systemd.services.ipfs =
+      let
+        # cidPinList is a function that generates a multiline string like this:
+        #
+        # ipfs pin add x &
+        # ipfs pin add y &
+        #
+        # Where x and y are the elements of pinContentIDList. This is then used
+        # in the postStart script for IPFS. This means content is pinned
+        # declaratively when IPFS starts, rather than having to execute the
+        # command `ipfs pin x` manually as a user.
+        cidPinList = lib.mkIf
+          (builtins.isList cfg.pinContentIDList)
+          (
+            lib.concatStringsSep
+              "\n"
+              (
+                map
+                  (x: "ipfs pin add ${x} &")
```

Comment:
`ipfs pin add` blocks on a content ID. If you give it a content ID that doesn't exist, nothing is queued, and it will never finish.

For example, if `a` does not exist, but `b` does, `b` will never be downloaded, as it will be blocked. Nothing is queued by the IPFS daemon, unless we fork `ipfs pin add`. This makes sense anyway, because if `a` does come into existence at any point, IPFS will succeed eventually, since it blocks until the CID exists.

```
ipfs pin add a b
```
If we fork it, everything works as expected. The behavior of `ipfs pin add` is idempotent.

Reply:
Maybe you want to create a `systemd` `oneshot` service **per** list item?

Reply:
I'm not sure how to implement that, would there not be a large CPU overhead for large lists then? I'm not sure about the implications of it.

Reply:
How large are we talking here?

I'm trying to think exactly what would happen if you `fork` off a bunch of processes in a shell script that is run in the `ExecStartPre`... ping @arianvp and @flokli who probably have smart opinions on this.

Reply:
Let's say you want to encode 100 CIDs to be pinned when IPFS starts up. That seems like a reasonable usage of this option.

Reply:
This is probably still fine - some of the [All Systems Go!](https://all-systems-go.io/) mentions they run A LOT of these units. Let's make each pin an individual unit, It's less costly.

Especially in the case of a lot of pins existing, and the addition of another one, we don't fork away all these `ipfs pin…` commands for stuff that's already pinned.